asQ/allatoncesystem.py

```diff
@@ -0,0 +1,503 @@
+
+        self.check_index(i, index_range=from_range)
+
+        # deal with -ve indices
+        i = i % self.max_indices[from_range]
+
+        # no shift needed
+        if to_range == from_range:
+            return i
+
+        # index of first timestep in slice
+        index0 = sum(self.time_partition[:self.time_rank])
+
+        if to_range == 'slice':  # 'from_range' == 'window'
+            i -= index0
+
+        if to_range == 'window':  # 'from_range' == 'slice'
+            i += index0
+
+        self.check_index(i, index_range=to_range)
+
+        return i
+
+    @PETSc.Log.EventDecorator()
+    def set_component(self, step, cpt, wnew, index_range='slice', f_alls=None):
+        '''
+        Set component of solution at a timestep to new value
+
+        :arg step: index of timestep
+        :arg cpt: index of component
+        :arg wout: new solution for timestep
+        :arg index_range: is index in window or slice?
+        :arg f_alls: an all-at-once function to set timestep in. If None, self.w_alls is used
+        '''
+        step_local = self.shift_index(step, from_range=index_range, to_range='slice')
+        self.check_index(cpt, index_range='component')
+
+        if f_alls is None:
+            f_alls = self.w_alls
+
+        # index of first component of this step
+        index0 = self.ncomponents*step_local
+
+        f_alls[index0 + cpt].assign(wnew)
+
+    @PETSc.Log.EventDecorator()
+    def get_component(self, step, cpt, index_range='slice', wout=None, name=None, f_alls=None, deepcopy=False):
+        '''
+        Get component of solution at a timestep
+
+        :arg step: index of timestep to get
+        :arg cpt: index of component
+        :arg index_range: is index in window or slice?
+        :arg wout: function to set to component (component returned if None)
+        :arg name: name of returned function if deepcopy=True. Ignored if wout is not None
+        :arg f_alls: an all-at-once function to get timestep from. If None, self.w_alls is used
+        :arg deepcopy: if True, new function is returned. If false, handle to component of f_alls is returned. Ignored if wout is not None
+        '''
+        step_local = self.shift_index(step, from_range=index_range, to_range='slice')
+        self.check_index(cpt, index_range='component')
+
+        if f_alls is None:
+            f_alls = self.w_alls
+
+        # index of first component of this step
+        index0 = self.ncomponents*step_local
+
+        # required component
+        wget = f_alls[index0 + cpt]
+
+        if wout is not None:
+            wout.assign(wget)
+            return wout
+
+        if deepcopy is False:
+            return wget
+        else:  # deepcopy is True
+            wreturn = fd.Function(self.function_space.sub(cpt), name=name)
+            wreturn.assign(wget)
+            return wreturn
+
+    @PETSc.Log.EventDecorator()
+    def set_timestep(self, step, wnew, index_range='slice', f_alls=None):
```

Comment:
can we call this set_field or something? because it doesn't set the tilmestep

Reply:
set_value_at_timestep

Reply:
Sure. Overloaded naming is confusing.
Do you have a preference between `set_field` and `set_value_at_timestep`? More concise or more explicit?